In [4]:
# Link to GitLab Repository
GitRep = "https://gitlab.gbar.dtu.dk/s224222/computational_assignment1"


In [5]:
import requests
from bs4 import BeautifulSoup
import difflib
import pandas as pd

Part I - WEBSCRAPING

In [6]:
url = "https://ic2s2-2023.org/program"

response = requests.get(url)
soup = BeautifulSoup(response.text, "html.parser")

In [7]:
soup.find_all("i")

[<i>Chair: Claudia Wagner</i>,
 <i><u>Jonas L Juul</u>, Jon Kleinberg</i>,
 <i><u>Chloe Ahn</u>, Xinyi Wang</i>,
 <i><u>Giuseppe Russo</u>, luca verginer, Manoel Horta Ribeiro, Giona Casiraghi</i>,
 <i><u>Almog Simchon</u>, Adam Sutton, Matthew Edwards, Stephan Lewandowsky</i>,
 <i><u>Arianna Pera</u>, Manuel Vimercati, Matteo Palmonari</i>,
 <i><u>Mohammed Alsobay</u>, Abdullah Almaatouq, David G. Rand, Duncan J. Watts</i>,
 <i>Sara Venturini, <u>Satyaki Sikdar</u>, Francesco Rinaldi, Francesco Tudisco, Santo Fortunato</i>,
 <i><u>Isabella Loaiza</u>, Takahiro Yabe, Alex Pentland</i>,
 <i>Chair: Taha Yasseri</i>,
 <i><u>Silvia De Sojo Caso</u>, Mia Ann Jørgensen, Sune Lehmann, Laura Alessandretti</i>,
 <i><u>Emil Bakkensen Johansen</u>, Mathias Wullum Nielsen</i>,
 <i><u>Rubén Rodríguez Casañ</u>, Antonio Ariño Villarroya</i>,
 <i><u>Sunny Rai</u>, Ashley Francisco, Salvatore Giorgi, Brenda Curtis, Lyle Ungar, Sharath Chandra Guntuku</i>,
 <i><u>Allison Koenecke</u>, Eric Giannella, R

In [8]:
# Get all the names into a table
names = []
for i in soup.find_all("i"):
    names.append(i.text)

names = ",".join(names)


# if "chair" in names: Remove the "chair"

names = names.replace("Chair: ", "")
names = names.split(",")
map(str.strip, names)
print(names)

['Claudia Wagner', 'Jonas L Juul', ' Jon Kleinberg', 'Chloe Ahn', ' Xinyi Wang', 'Giuseppe Russo', ' luca verginer', ' Manoel Horta Ribeiro', ' Giona Casiraghi', 'Almog Simchon', ' Adam Sutton', ' Matthew Edwards', ' Stephan Lewandowsky', 'Arianna Pera', ' Manuel Vimercati', ' Matteo Palmonari', 'Mohammed Alsobay', ' Abdullah Almaatouq', ' David G. Rand', ' Duncan J. Watts', 'Sara Venturini', ' Satyaki Sikdar', ' Francesco Rinaldi', ' Francesco Tudisco', ' Santo Fortunato', 'Isabella Loaiza', ' Takahiro Yabe', ' Alex Pentland', 'Taha Yasseri', 'Silvia De Sojo Caso', ' Mia Ann Jørgensen', ' Sune Lehmann', ' Laura Alessandretti', 'Emil Bakkensen Johansen', ' Mathias Wullum Nielsen', 'Rubén Rodríguez Casañ', ' Antonio Ariño Villarroya', 'Sunny Rai', ' Ashley Francisco', ' Salvatore Giorgi', ' Brenda Curtis', ' Lyle Ungar', ' Sharath Chandra Guntuku', 'Allison Koenecke', ' Eric Giannella', ' Robb Willer', ' Sharad Goel', 'Ziv Epstein', ' Hause Lin', 'Levin Brinkmann', ' Bramantyo Supriyatn

In [9]:
# Remove duplicates, strip and lower case
for i in range(len(names)):
    names[i] = names[i].lower()
    names[i] = names[i].strip()
names = list(set(names))
# and sort alphabetically

names.sort()
print(names)

['aaron clauset', 'aaron j. schwartz', 'aaron schein', 'aaron smith', 'abbas haidar', 'abby smith', 'abdulkadir celikkanat', 'abdullah almaatouq', 'abdullah zameek', 'abeer elbahrawy', 'adam finnemann', 'adam frank', 'adam h. russell', 'adam stefkovics', 'adam sutton', 'aditi dutta', 'adriano belisario', 'adrienne mendrik', 'agnieszka czaplicka', 'agnieszka falenska', 'aguru ishibashi', 'ahmad hesam', 'ahmed nasser mostafa', 'aidan combs', 'aidar zinnatullin', 'akeela careem', 'akhil arora', 'akira hashimoto', 'akira matsui', 'akira tsurushima', 'akrati saxena', 'alain barrat', 'alan paul kwan', 'alba motes rodrigo', 'albert-laszlo barabasi', 'alberto amaduzzi', 'alejandro beltran', 'alejandro dinkelberg', 'alejandro hermida carrillo', 'aleksandra urman', 'alessandra urbinati', 'alessandro de gaetano', 'alessandro flamini', 'alessandro flammini', 'alessandro gambetti', 'alessandro lomi', 'alessia antelmi', 'alessia melegaro', 'alessio vincenzo cardillo', 'alex mielke', 'alex pentland',

In [10]:
# Remove duplicates where the names are very similar using sequence matcher

duplicate = []
for i in range(len(names)):
    for j in range(i+1, len(names)):
        if difflib.SequenceMatcher(None, names[i], names[j]).ratio() > 0.95:
            duplicate.append(names[j].lower().strip())

print(duplicate)

['alessandro flammini', 'anne c. kroon', 'diogo pacheco', 'duncan j. watts', 'fabio carrella', 'federico barrera-lemarchand', 'maximilian schich', 'pantelis p. analytis', 'scott a. hale', 'sonja m schmer-galunder']


In [11]:
# Remove one of the duplicates from the list
names = [x for x in names if x not in duplicate]

In [14]:
print(len(names))

1471


We get 1471 authors after having cleaned the data

In [13]:
# Create a txt file with the names

with open("names.txt", "w") as f:
    for name in names:
        f.write(name + "\n")



Explain the process you followed to web-scrape the page. Which choices did you make to accurately retreive as many names as possible? Which strategies did you use to assess the quality of your final list? Explain your reasoning and your choices (answer in max 150 words).

THE PROCESS:  (PERHAPS MORE HERE)

In order to retrieve as many names as possible, we ...

To clean the data, we lowercased, stripped and removed obvious duplicates. To take human error into account, we used a sequence matcher from the difflib library where we compared each name and if they were 95% or more similar, we removed them. We checked each of them by printing the two similar names, and thereby, trying to ensure we didn't remove any that could be two seperate people. 

PART II - READY MADE VS. CUSTOM MADE


    What are pros and cons of the custom-made data used in Centola's experiment (the first study presented in the lecture) and the ready-made data used in Nicolaides's study (the second study presented in the lecture)? You can support your arguments based on the content of the lecture and the information you read in Chapter 2.3 of the book (answer in max 150 words).
    How do you think these differences can influence the interpretation of the results in each study? (answer in max 150 words)


ANSWER HERE

PART III - GATHERING RESEARCH ARTICLES USING OPENALEX API

This will need some work.... I only made it a liiiiiitle bit of the way

In [ ]:
authors = pd.read_csv("IC2S2authors.csv")

In [ ]:
# Created to filter what papers we actually want to look at
fields_to_ids = {'Sociology': 'https://openalex.org/C144024400',
                 'Psychology': 'https://openalex.org/C15744967',
                 'Economics': 'https://openalex.org/C162324750',
                 'Political Science': 'https://openalex.org/C17744445',
                 'Mathematics': 'https://openalex.org/C33923547}',
                 'Physics':'https://openalex.org/C121332964',
                 'Computer Science': 'https://openalex.org/C41008148'
                }


In [ ]:
### Showing the basic structure of the API request
base_url = 'https://api.openalex.org/' 
resource = 'works'
filterstring = ["cited_by_count:>10",
                 "authors_count:<10", 
                 f"concepts.id:{fields_to_ids['Sociology']}|{fields_to_ids['Psychology']}|{fields_to_ids['Economics']}|{fields_to_ids['Political Science']}", 
                 f"concepts.id:{fields_to_ids['Mathematics']}|{fields_to_ids['Physics']}|{fields_to_ids['Computer Science']}"]
filterstring = ",".join(filterstring)
request = {
    "filter": filterstring,
    "cursor": '*',
    'mailto': 's224222@dtu.dk'
}
# This may be changed to a different email... idk - Sophia

In [ ]:
rq = requests.get(base_url + resource, params=request)
data = rq.json()

In [ ]:
data['meta']
# I don't know what this is... - Sophia 

In [ ]:
# Ensuring we keep the authors only within limits
authors = authors[(authors['works_count'] > 5) & (authors['works_count'] < 5000)]


In [ ]:
author_ids = authors['id'].values, len(authors)

Don't know how to actually continue from here... to get the correct result so to speak - Sophia

PART IV - THE NETWORK OF COMPUTATIONAL SCIENTISTS